### ISSUE TITLE
Environment Variables Not Available

### ISSUE THREAD

**Issue by clwehunt** [paratestphp/paratest#721]
| Q | A
| --- | ---
| ParaTest version | 6.8.1
| PHPUnit version | 9.5.28
| PHP version | 7.4.30

#### Summary

I am testing out a very simple phpUnit test.

```php
public function testAppEnvVariableIsTest()
    {
        $this->assertEquals($_ENV['APP_ENV'], 'test');
    }
```

#### Current behavior

Running phpunit the results are:
Environment: test
02/02/2023 11:25:02 pm

```
PHPUnit 9.5.28 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.4.30
Configuration: /phpunit.xml

.....                                                               5 / 5 (100%)

Time: 00:00.019, Memory: 6.00 MB

OK (5 tests, 5 assertions)
```

Running paratest via the following command: `vendor/bin/paratest --runner WrapperRunner`

```
ParaTest v6.8.1 upon PHPUnit 9.5.28 by Sebastian Bergmann and contributors.

Environment: test
02/02/2023 11:26:08 pm
...E.                                                               5 / 5 (100%)

Time: 00:00.464, Memory: 6.00 MB

There was 1 error:

1) ExampleTest::testAppEnvVariableIsTest
Undefined index: APP_ENV

FAILURES!                          
Tests: 5, Assertions: 4, Errors: 1.
```

You can see from the output above that paratest loads the same bootstrap file and can access it as the environment is printed correctly.

#### How to reproduce: command, code and error stack trace

Composer.json require:
```
"phpunit/phpunit": "^9.5",
"brianium/paratest": "^6.8",
"vlucas/phpdotenv": "^5.5",
```

Have an .env file with APP_ENV="test"

Have a bootstrap file referenced in phpunit.xml that loads the .env file:
```php
require 'vendor/autoload.php';
$dotenv = Dotenv\Dotenv::createImmutable(__DIR__);
$dotenv->load();
```

Have a test defined as:
```php
public function testAppEnvVariableIsTest()
    {
        $this->assertEquals($_ENV['APP_ENV'], 'test');
    }
```

Run vendor/bin/paratest --runner WrapperRunner

#### Expected behavior

I would expect the output to be the same as phpUnit, the environment variable to be available, and the assertion to be true:
ParaTest v6.8.1 upon PHPUnit 9.5.28 by Sebastian Bergmann and contributors.

Environment: test
02/02/2023 11:28:15 pm
```
....                                                                5 / 5 (100%)

Time: 00:00.429, Memory: 6.00 MB

OK (5 tests, 5 assertions)
```

**Comment by steltner** [comment]
Had the same problem, cost me a day of my life, the paratest worker has every ENV-parameter as $_SERVER parameter (I don't know why) and the ServerConstAdapter we use as default adapter by creating Dotenv with Dotenv::createImmutable(__DIR__) will return that every ENV-parameter is externally defined (see ImmutableWriter->isExternallyDefined($parameterName) and so it will not be loaded from .env and you have no $_ENV parameters in a worker.

You can bypass this problem by not loading the adapter ServerConstAdapter:
```php
        $repository = RepositoryBuilder::createWithNoAdapters()
            ->addAdapter(EnvConstAdapter::class)
            ->immutable()
            ->make();

        $dotenv = Dotenv::create($repository, __DIR__);
        $dotenv->load();
```

**Comment by steltner** [comment]
Had the same problem, cost me a day of my life, the paratest worker has every ENV-parameter as $_SERVER parameter (I don't know why) and the ServerConstAdapter we use as default adapter by creating Dotenv with `Dotenv::createImmutable(__DIR__)` will return that every ENV-parameter is externally defined (see `ImmutableWriter->isExternallyDefined($parameterName)`) and so it will not be loaded from .env and you have no $_ENV parameters in a worker.

You can bypass this problem by not loading the adapter ServerConstAdapter:
```php
        $repository = RepositoryBuilder::createWithNoAdapters()
            ->addAdapter(EnvConstAdapter::class)
            ->immutable()
            ->make();

        $dotenv = Dotenv::create($repository, __DIR__);
        $dotenv->load();
```

**Comment by Slamdunk** [comment]
@steltner Would you be able to write a brief paragraph about this issue and propose a PR with it?

**Comment by clwehunt** [comment]
> @steltner Would you be able to write a brief paragraph about this issue and propose a PR with it?

@steltner @Slamdunk Just following up on this issue if you have time.  Thanks

**Comment by clwehunt** [comment]
> Had the same problem, cost me a day of my life, the paratest worker has every ENV-parameter as $_SERVER parameter (I don't know why) and the ServerConstAdapter we use as default adapter by creating Dotenv with `Dotenv::createImmutable(__DIR__)` will return that every ENV-parameter is externally defined (see `ImmutableWriter->isExternallyDefined($parameterName)`) and so it will not be loaded from .env and you have no $_ENV parameters in a worker.
> 
> You can bypass this problem by not loading the adapter ServerConstAdapter:
> 
> ```
>         $repository = RepositoryBuilder::createWithNoAdapters()
>             ->addAdapter(EnvConstAdapter::class)
>             ->immutable()
>             ->make();
> 
>         $dotenv = Dotenv::create($repository, __DIR__);
>         $dotenv->load();
> ```

@steltner Thanks so much, In the meantime I have this workaround and it seems to be working ok.

**Comment by leonardoldr23** [comment]
Any update on this? Due to the recent upgrade on Laravel and using Pest this behavior breaks our CI.

I would lovely send a PR with the fix, but unfortunately I really don't know even where to start here.

**Comment by Slamdunk** [comment]
> but unfortunately I really don't know even where to start here

Without being able to reproduce the big, me neither

**Comment by leonardoldr23** [comment]
At least the use case I can provide.

- Install a fresh laravel project
- install Pestphp/pest on the project
- install the laravel plugin for pest (just for a shortcut)
- Write a very simple test case (  php artisan pest:test Example --unit )
- Execute it with ( ./vendor/bin/pest --filter=Example --colors=never --compact ) and it will be without formatting
- Execute with parallel (  ./vendor/bin/pest --filter=Example --colors=never --parallel --processes=1 ) and it will output all formatted as if the --colors option was never passed.

As it works from the common run and not in parallel, we assumed that might be a problem here, but can also be a problem on Pest or phpunit itself.

**Comment by Slamdunk** [comment]
@leonardoldr23 can you create a dedicated repository with the minimal code needed for the bug to raise, and instructions on how to expose it?

**Comment by leonardoldr23** [comment]
@Slamdunk of course, here it is: https://github.com/leonardoldr23/envparatestshowcase

To run, clone/download then:

- composer install (I'm using version 2.4.1)

Afterwards, to reproduce the problem simply run this commands and you will see the issue:

- php artisan test   (first, without paratest and with colors)
- php artisan test --colors=never  (then to see that the flag is working right in pest/laravel)
- php artisan test --parallel  (now parallel, no flags at all)
- php artisan test --parallel --colors=never  (parallel and the flag, still not working).

I will send in attachment an image on the problem happening too:
![paratest-issue](https://github.com/paratestphp/paratest/assets/23258783/e9104ab6-b756-4334-8a31-1699fa385e34)

As you can see, when using parallel the output keeps the ansii characters which are very bad on Jenkins (yeah, Jenkins should support them too :D)

**Comment by Slamdunk** [comment]
How are CLI flags associated with ENV variables?

**Comment by leonardoldr23** [comment]
They are passed as $_SERVER variables just as like environments after the reading.

**Comment by Slamdunk** [comment]
Then that sounds unrelated to ParaTest, which never touches `$_SERVER` variable

**Comment by leonardoldr23** [comment]
I don't remember why I came to relate with this issue exactly, but I showed the issue running at all, if you need I can open a new issue

**Comment by Slamdunk** [comment]
You showed the issue, but you didn't prove it's ParaTest fault and not, lets say, Pest or Laravel

**Comment by leonardoldr23** [comment]
Ok, I really don't know how can I do it in any other way than showing that without paratest it works, and with paratests it not works. Just leave it be then, we will find another way or tool to parallelize the tests. Thanks for your time.

**Comment by Slamdunk** [comment]
> showing that without paratest it works, and with paratests it not works

Pest overwrites many classes of ParaTest, so when you activate the `--parallel` flag it's not just ParaTest alone involved in the output. I suggest you to raise this issue in Pest.
